### DISCUSSION TITLE
how do I compare JSON to the 'null' JSON string?

### DISCUSSION THREAD

**Discussion by ljmc-github** [sqlalchemy/sqlalchemy#11394] [Usage Questions] [answered]
I think I may have found a bug (or maybe undefined behaviour that is working against me) wrt to `null` values in JSONB columns.

When inserting into the following table, as expected a `'null'::JSONB` is passed for Alice's `optional_data`, however, when querying Alice back, the filtering produces a `optional_data IS NULL` which does not work for round tripping Alice.

```python
"""
CREATE TABLE null_sql_json (
    id BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
    name TEXT NOT NULL,
    optional_data JSONB
);
"""

from sqlalchemy import MetaData, Table, create_engine, select
from sqlalchemy.orm import Session, declarative_base, registry

engine = create_engine(
    "postgresql+psycopg2://postgres:postgres@localhost:5432/postgres",
    echo=True,
)

metadata = MetaData()
mapper_registry = registry()

class NullSqlJson: ...

mapper_registry.map_imperatively(
    NullSqlJson, Table("null_sql_json", metadata, autoload_with=engine)
)

with Session(engine) as session:
    session.add(NullSqlJson(name="Alice", optional_data=None))
    session.commit()
# emits
# INSERT INTO null_sql_json (name, optional_data) VALUES (%(name)s, %(optional_data)s) RETURNING null_sql_json.id
# {'name': 'Alice', 'optional_data': 'null'}

with Session(engine) as session:
    print(
        session.execute(
            select(NullSqlJson).filter_by(name="Alice", optional_data=None)  # .filter produces same output
        ).scalar_one_or_none()
    )
# emits
# SELECT null_sql_json.id, null_sql_json.name, null_sql_json.optional_data
# FROM null_sql_json
# WHERE null_sql_json.name = %(name_1)s AND null_sql_json.optional_data IS NULL
# {'name_1': 'Alice'}
# prints None
```

Software versions:
- MacOS 14.4.1 (23E224)
- Python 3.11.9 (from pyenv)
- psycopg2-binary 2.9.9
- SQLAlchemy 2.0.30
- PostgreSQL 16.2 (Debian 16.2-1.pgdg120+2) on x86_64-pc-linux-gnu, compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit

**NB.** this obviously uses a reflected table, it is an adaptation of the code where I found the problem. I am happy to work on a declarative version if that would help.

**Comment by zzzeek** [comment, marked as answer]
see the section "Support for JSON null vs. SQL NULL" at https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.JSON .   for this comparison you want to use the [JSON.NULL ](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.JSON.NULL).

**Comment by ljmc-github** [comment]
thank you for the fast reply, I suspected there would be something like that, I settled on using the `column_reflect` event and retype the column with `JSONB(none_as_null=True)` to always have SQL NULL.

I'm not sure how operator override is done, but would it not make sense to catch the JSON columns so `None` is interpreted the same between `DQL` and `DML` ? I understand the `== None` defaulting to `IS NULL`, but can it not be overriden for JSON (and subclasses) ?

**Reply by zzzeek** [reply]
I think JSON comparison in SQL is an unusual case to start with, since there are many equivalent JSON strings that don't compare as equal at the string level, so I think people usually would expect "col == None" to mean IS NULL even in a JSON context where the 'null' value is persisted.

operator overrides are illustrated at https://docs.sqlalchemy.org/en/20/core/custom_types.html#redefining-and-creating-new-operators, here is a demo for what you are seeking

```py
from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import JSON
from sqlalchemy import select
from sqlalchemy.orm import declarative_base
Base = declarative_base()


class MyJSON(JSON):
    class comparator_factory(JSON.Comparator):
        def __eq__(self, other):
            if other is None:
                other = JSON.NULL
            return super().__eq__(other)


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(MyJSON)

print(select(A).where(A.data == None))
```

**Reply by ljmc-github** [reply]
I agree it's a unusual one, and that's why I've reverted to None as SQL NULL which serves my purposes very well, i.e. make DML and DQL look the same.

Thank you for demo of comparator factory, very interesting, SQLAlchemy just keeps on giving.
